google-ads-examples/src/main/java/com/google/ads/googleads/examples/advancedoperations/AddRealEstateFeed.java

```diff
@@ -0,0 +1,507 @@
+   * @param feedResourceName the feed resource name.
+   */
+  private void createFeedItems(
+      GoogleAdsClient googleAdsClient,
+      long customerId,
+      Map<RealEstatePlaceholderField, FeedAttribute> feedAttributes,
+      String feedResourceName) {
+
+    // Creates the listing ID feed attribute value.
+    FeedItemAttributeValue listingId =
+        FeedItemAttributeValue.newBuilder()
+            .setFeedAttributeId(
+                Int64Value.of(
+                    feedAttributes.get(RealEstatePlaceholderField.LISTING_ID).getId().getValue()))
+            .setStringValue(StringValue.of("ABC123DEF"))
+            .build();
+    // Creates the listing name feed attribute value.
+    FeedItemAttributeValue listingName =
+        FeedItemAttributeValue.newBuilder()
+            .setFeedAttributeId(
+                Int64Value.of(
+                    feedAttributes.get(RealEstatePlaceholderField.LISTING_NAME).getId().getValue()))
+            .setStringValue(StringValue.of("Two bedroom with magnificent views"))
+            .build();
+    // Creates the final URLs feed attribute value.
+    FeedItemAttributeValue finalUrls =
+        FeedItemAttributeValue.newBuilder()
+            .setFeedAttributeId(
+                Int64Value.of(
+                    feedAttributes.get(RealEstatePlaceholderField.FINAL_URLS).getId().getValue()))
+            .addStringValues(StringValue.of("http://www.example.com/listings/"))
+            .build();
+
+    // Optionally insert additional attributes here, such as address, city, description, etc.
+
+    // Creates the image URL feed attribute value.
+    FeedItemAttributeValue imageUrl =
+        FeedItemAttributeValue.newBuilder()
+            .setFeedAttributeId(
+                Int64Value.of(
+                    feedAttributes.get(RealEstatePlaceholderField.IMAGE_URL).getId().getValue()))
+            .setStringValue(
+                StringValue.of("http://www.example.com/listings/images?listing_id=ABC123DEF"))
+            .build();
+    // Creates the contextual keywords feed attribute value.
+    FeedItemAttributeValue finalUrl =
+        FeedItemAttributeValue.newBuilder()
+            .setFeedAttributeId(
+                Int64Value.of(
+                    feedAttributes
+                        .get(RealEstatePlaceholderField.CONTEXTUAL_KEYWORDS)
+                        .getId()
+                        .getValue()))
+            .addStringValues(StringValue.of("beach community"))
+            .addStringValues(StringValue.of("ocean view"))
+            .addStringValues(StringValue.of("two bedroom"))
+            .build();
+
+    // Creates the FeedItem, specifying the Feed ID and the attributes created above.
+    FeedItem feedItem =
+        FeedItem.newBuilder()
+            .setFeed(StringValue.of(feedResourceName))
+            .addAttributeValues(listingId)
+            .addAttributeValues(listingName)
+            .addAttributeValues(finalUrls)
```

Comment:
Do we need finalUrls and finalUrl? Surely just one of these would suffice?

Reply:
I did it this way to mirror the example in the previous Dynamic Remarketing guide, but I think it's helpful to see an attribute that is a list type anyhow.